nautilus_trader/adapters/interactive_brokers/client/contract.py

```diff
@@ -0,0 +1,215 @@
+# -------------------------------------------------------------------------------------------------
+#  Copyright (C) 2015-2021 Nautech Systems Pty Ltd. All rights reserved.
+#  https://nautechsystems.io
+#
+#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
+#  You may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+# -------------------------------------------------------------------------------------------------
+
+import functools
+from typing import TYPE_CHECKING, Any
+
+from ibapi.common import SetOfFloat
+from ibapi.common import SetOfString
+from ibapi.contract import ContractDetails
+from ibapi.utils import current_fn_name
+
+from nautilus_trader.adapters.interactive_brokers.common import IBContract
+from nautilus_trader.adapters.interactive_brokers.common import IBContractDetails
+
+
+if TYPE_CHECKING:
```

Comment:
Interesting why we need conditional `TYPE_CHECKING` here?

Reply:
It's to avoid a circular reference -- `contract.py` needs to import the `InteractiveBrokersClient` from `client.py` but `client.py` needs to import the `InteractiveBrokersContractManager` from `contract.py`.

Reply:
Could this be solved by changing the structure of the subpackage and shuffling some modules instead?

Reply:
I'm not sure that this would be easy to do due to the way IB's `EClient` and `EWrapper` classes work, but if you or @rsmb7z have thoughts, I'm open ears and happy to incorporate the feedback. Here's my thinking / understanding:

The reason this is tricky is because the `InteractiveBrokersClient` is the class that holds the connection to the TWS API (via the `EWrapper` (to receive messages from the TWS API via websocket) and `EClient` (to send messages to the TWS API via websocket). All the new subcomponents, which are almost entirely comprised of methods simply pulled out of the `Client` to reduce the class size, need to send messages to the TWS API. As a result, the `Client` is passed to the subcomponents as a parameter, but as noted in an earlier comment, the `Client` also instantiates the new subcomponents, so it can delegate to them.

This is a circular dependency, which means that there's a high level of coupling between classes, but in this case, I think it's hard to avoid.

Here is a simplified / contrived example that illustrates what's going on:

```
class InteractiveBrokersClient(Component, EWrapper):
    def __init__(self, ...):
        self._eclient = EClient(wrapper=self)
        self.market_data_manager = InteractiveBrokersMarketDataManager(self)

    # EWrapper override to receive historical bars
    def historicalData(self):
        # delegate to the subcomponent
        self.market_data_manager.receive_bars()

class InteractiveBrokersMarketDataManager:
    def __init__(self, client: InteractiveBrokersClient):
        self._client = client

    def request_bars(self, ...):
        return self._client._eclient.reqHistoricalData(...)

    def receive_bars(self):
        # do stuff then ask for more bars
        self.request_bars(...)
```

I'm going to continue to think about whether there is a way to decouple the dependency.

Reply:
I agree with your suggestion -- mixins are better in this case. I refactored the new subcomponent classes into mixins, and I think this eliminates the circular dependencies and is easier to understand.